omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+static void myInitHookFunction(initHookState state)
+{
+  switch(state) {
+    case initHookAfterIocRunning:
+      iocStarted = true;
+      break;
+    default:
+      break;
+  }
+}
+
+drvOmronEIP::drvOmronEIP(const char *portName,
+                          char *gateway,
+                          char *path,
+                          char *plcType,
+                          int debugLevel)
+
+    : asynPortDriver(portName,
+                     1,                                                                                                                                                  /* maxAddr */
+                     asynInt32Mask | asynUInt32DigitalMask | asynInt64Mask | asynFloat64Mask | asynInt32ArrayMask | asynOctetMask | asynDrvUserMask | asynInt8ArrayMask, /* Interface mask */
+                     asynInt32Mask | asynUInt32DigitalMask | asynInt64Mask | asynFloat64Mask | asynInt32ArrayMask | asynOctetMask | asynInt8ArrayMask,                   /* Interrupt mask */
+                     ASYN_CANBLOCK | ASYN_MULTIDEVICE,                                                                                                                   /* asynFlags */
+                     1,                                                                                                                                                  /* Autoconnect */
+                     0,                                                                                                                                                  /* Default priority */
+                     0),                                                                                                                                                 /* Default stack size*/
+      initialized_(false),
+      omronExiting_(false)
+
+{
+  static const char *functionName = "drvOmronEIP";
+  tagConnectionString_ = "protocol=ab-eip&gateway="+(std::string)gateway+"&path="+(std::string)path+"&plc="+(std::string)plcType;
+  std::cout<<"Starting driver with connection string: "<< tagConnectionString_<<std::endl;;
+  epicsAtExit(omronExitCallback, this);
+  plc_tag_set_debug_level(debugLevel);
+  initHookRegister(myInitHookFunction);
+  asynStatus status = (asynStatus)(epicsThreadCreate("optimiseTags",
+                          epicsThreadPriorityMedium,
+                          epicsThreadGetStackSize(epicsThreadStackMedium),
+                          (EPICSTHREADFUNC)optimiseTagsC,
+                          this) == NULL);
+  if (status!=0){
+    std::cout<< "Error, Driver initiation returned asyn status:" << status << std::endl;
+  }
+  initialized_ = true;
+}
+
+asynStatus drvOmronEIP::createPoller(const char *portName, const char *pollerName, double updateRate)
+{
+  int status;
+  omronEIPPoller* pPoller = new omronEIPPoller(portName, pollerName, updateRate);
+  pollerList_[pPoller->pollerName_] = pPoller;
+  status = (epicsThreadCreate(pPoller->pollerName_,
```

Comment:
There is also an epicsThread C++ interface which might make this code look a little bit more like "proper" C++ ;-) 

It just wraps the C interface and so there is no real technical reason to change. Just aestethics if you're into that...

Reply:
I did not know about this, I will try and remember for the next time I need to use epics threading.